crates/bevy_ecs/src/schedule/state.rs

```diff
@@ -1,260 +1,590 @@
+                            state.stack[state.stack.len() - 2].clone(),
+                        ));
+                    }
+                }
             }
         }
+        Some(ScheduledOperation::Push(next)) => {
+            let last_type_id = state.stack.last().unwrap().clone();
+            state.transition = Some(StateTransition::Pausing(last_type_id, next));
+        }
+        Some(ScheduledOperation::Pop) => {
+            state.transition = Some(StateTransition::ExitingToResume(
+                state.stack[state.stack.len() - 1].clone(),
+                state.stack[state.stack.len() - 2].clone(),
+            ));
+        }
+        None => match state.transition.take() {
+            Some(StateTransition::ExitingFull(p, n)) => {
+                state.transition = Some(StateTransition::Entering(p, n.clone()));
+                state.stack[0] = n;
+            }
+            Some(StateTransition::Pausing(p, n)) => {
+                state.transition = Some(StateTransition::Entering(p, n.clone()));
+                state.stack.push(n);
+            }
+            Some(StateTransition::ExitingToResume(p, n)) => {
+                state.stack.pop();
+                state.transition = Some(StateTransition::Resuming(p, n));
+            }
+            Some(StateTransition::PreStartup) => {
+                state.transition = Some(StateTransition::Startup);
+            }
+            _ => {}
+        },
+    };
+    if state.transition.is_none() {
+        *prep_exit = true;
     }
+
+    ShouldRun::YesAndCheckAgain
 }
 
-impl<T: Clone> Deref for State<T> {
-    type Target = T;
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::prelude::*;
+
+    #[derive(Copy, Clone, Eq, PartialEq, Debug)]
+    enum MyState {
+        S1,
+        S2,
+        S3,
+        S4,
+        S5,
+        S6,
+        Final,
+    }
 
-    fn deref(&self) -> &Self::Target {
-        &self.current
+    #[test]
+    fn state_test() {
+        let mut world = World::default();
+
+        world.insert_resource(Vec::<&'static str>::new());
+        world.insert_resource(State::new(MyState::S1));
+
+        let mut stage = SystemStage::parallel();
+
+        stage.add_system_set(State::<MyState>::make_driver());
+        stage
+            .add_system_set(
+                State::on_enter_set(MyState::S1).with_system(
+                    (|mut r: ResMut<Vec<&'static str>>| r.push("startup"))
+                        .system()
+                        .label("0"),
+                ),
+            )
+            .add_system_set(
+                State::on_update_set(MyState::S1).with_system(
+                    (|mut r: ResMut<Vec<&'static str>>, mut s: ResMut<State<MyState>>| {
+                        r.push("update S1");
+                        s.overwrite_next(MyState::S2).unwrap();
+                    })
+                    .system()
+                    .label("1")
+                    .before("0"),
```

Comment:
I think we shouldn't include ordering in the tests / the implementation shouldn't require it. The proper execution order should be enforced by the state lifecycle (and the tests should test for that). When I remove the ordering, it seems like this generally already works, but the last few states/transitions are non-deterministic?

Reply:
It makes sense that multiple "inactive" states aren't ordered, so maybe we only need to insert explicit ordering between them (for the sake of consistent tests).